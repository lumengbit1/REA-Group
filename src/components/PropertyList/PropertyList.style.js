import styled, { css } from 'styled-components';

const StyledButton = css`
  position: absolute;
  left: 50%;
  bottom: 1%;
  transform: translate(-50%, -50%);
  width: 11.25rem;
  height: 1.25rem;
  border-radius: 0.625rem;
  display: none;

  &:hover {
    cursor: pointer;
  }
`;

export const RootContainer = styled.div`
  display: flex;
  width: 100%;
`;

export const Button = styled.button`
  ${StyledButton};

  background-color: #f3c1c2;
  border: 0.0625rem solid #cc0007;
  color: #a21519;

  ${(props) => props.isResult && css`
    background-color: #e3f1df;
    border: 0.0625rem solid #4d8437;
    color: #4b8333;
  `}
`;

export const PropertyContainer = styled.div`
  display: flex;
  flex-direction: column;
  position: relative;
  width: 12.5rem;
  height: 11.25rem;
  border: 0.0625rem solid #000;
  border-radius: 0.625rem;
  margin: 0.625rem;

  &:hover {
    button {
      display: block;
    }
  }
`;

export const Title = styled.div`
  font-size: 1.875rem;
  font-weight: bold;
  margin: 0.625rem;
`;

export const Container = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  width: ${(props) => `${100 / props.length}%`};
`;
