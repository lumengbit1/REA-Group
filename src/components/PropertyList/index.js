import React, { useEffect } from 'react';
import { inject } from 'mobx-react';
import { observer } from 'mobx-react-lite';
import CSSModules from 'react-css-modules';
import styles from './propertyList.less';
import Loadable from 'react-loadable';

function PropertyList({ rootStore, area, btnText, btnClass }) {
    const propertyData = area === 'results' ? rootStore.formStore.resultsData : rootStore.formStore.savedData;
    const LoadableProperty = Loadable({
        loader: () => import('./../Property'),
        loading() {
            return <>Loading...</>;
        }
    });

    useEffect(() => {
        rootStore.formStore.getData();
    }, []);

    function onClick(id) {
        area === 'results' ? rootStore.formStore.addProperty(id) : rootStore.formStore.removeProperty(id);
    }

    return (
        <>
            {propertyData.map(item => {
                return (
                    <div styleName="property">
                        <LoadableProperty
                            price={item.price}
                            color={item.agency.brandingColors.primary}
                            logo={item.agency.logo}
                            id={item.id}
                            mainImage={item.mainImage}
                        />
                        <button
                            data-testid="test"
                            styleName={btnClass}
                            onClick={() => {
                                onClick(item.id);
                            }}
                        >
                            {btnText}
                        </button>
                    </div>
                );
            })}
        </>
    );
}
export default inject('rootStore')(observer(CSSModules(PropertyList, styles)));
