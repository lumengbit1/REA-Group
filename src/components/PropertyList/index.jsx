import React from 'react';
import PropTypes from 'prop-types';
import ImmutablePropTypes from 'react-immutable-proptypes';
import withComponentLoading from './util';
import Property from '../Property';
import {
  Button,
  PropertyContainer,
} from './PropertyList.style';

const PropertyList = (props) => {
  const { records, btnText, isResult, clickAction, type } = props;

  return (
    <>
      {records.map((item) => (
        <PropertyContainer key={item.id} data-testid="property">
          <Property
            id="property"
            price={item.price}
            color={item.agency.brandingColors.primary}
            logo={item.agency.logo}
            mainImage={item.mainImage}
          />
          <Button
            data-testid={`test${type}`}
            isResult={isResult}
            onClick={() => clickAction(item.id)}
          >
            {btnText}
          </Button>
        </PropertyContainer>
      ))}
    </>
  );
};

PropertyList.propTypes = {
  records: ImmutablePropTypes.list.isRequired,
  btnText: PropTypes.string.isRequired,
  isResult: PropTypes.bool.isRequired,
  clickAction: PropTypes.func.isRequired,
  type: PropTypes.string.isRequired,
};

export default withComponentLoading(PropertyList);
