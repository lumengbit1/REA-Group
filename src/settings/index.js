export default {
  LOCATIONIQ_BASE_API_DOMAIN: 'https://eu1.locationiq.com/v1/search.php',
  OPEN_WEATHER_MAP_BASE_API_DOMAIN: 'https://api.openweathermap.org/data/2.5/onecall',
  OPEN_WEATHER_MAP_KEY: '3ad086fd12d0afd2bb6e33faffbea059',
  LOCATIONIQ_KEY: 'pk.2b73038ba628901801e4e1e93d258e7a',
};
