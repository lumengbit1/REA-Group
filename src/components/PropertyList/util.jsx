import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { map, isEmpty, size } from 'lodash';
import { getAction, add_property, remove_property } from '../../reducers/actions';
import settings from '../../settings';
import { RootContainer, Container, Title } from './PropertyList.style';

function componentType (type) {
  if (type === 'results') {
    return {
      clickAction: add_property,
      btnText: 'Add Property',
      isResult: true,
    };
  }

  return {
    clickAction: remove_property,
    btnText: 'Remove Property',
    isResult: false,
  };
}

const withComponentLoading = (WrappedComponent) => {
  function WithWrap () {
    const dispatch = useDispatch();

    const value = useSelector((state) => state.reducer.entities);
    const result = useSelector((state) => state.reducer.result);
    const loading = useSelector((state) => state.reducer.loading);

    React.useEffect(() => {
      dispatch(getAction(settings.BASE_API_DOMAIN));
    }, []);

    if (loading) return <div data-testid="loading">Loading...</div>;

    if (isEmpty(value)) return null;

    return (
      <RootContainer>
        {map(result, (item, index) => (
          <Container
            key={index}
            length={size(result) || 1}
          >
            <Title>
              {index}
            </Title>
            {map(item, (val) => (
              <WrappedComponent
                key={value[val].id}
                records={value[val]}
                isResult={componentType(index).isResult}
                clickAction={(params) => dispatch(componentType(index).clickAction(params))}
                btnText={componentType(index).btnText}
              />
            ))}
          </Container>
        ))}
      </RootContainer>
    );
  }

  return WithWrap;
};

export default withComponentLoading;
