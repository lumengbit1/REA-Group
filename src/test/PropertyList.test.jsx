import React from 'react';
import 'jest-styled-components';
import '@testing-library/jest-dom';
import * as redux from 'react-redux';

import { rest } from 'msw';
import { setupServer } from 'msw/node';

import { render, cleanup, waitFor, fireEvent } from '@testing-library/react';
import thunk from 'redux-thunk';
import { createStore, applyMiddleware } from 'redux';
import { combineReducers } from 'redux-immutable';
import PropertyList from '../components/PropertyList';
import eventList from '../reducers/reducer';
import settings from '../settings';

const store = createStore(combineReducers({ value: eventList }), applyMiddleware(thunk));

const mockResultsData = [
  {
    price: '$726,500',
    agency: {
      brandingColors: {
        primary: '#ffe512',
      },
      logo: 'https://i1.au.reastatic.net/170x32/d9e65c666e427e655fb63dcfe73f50d4ac7ff9a58c173db9474bd92e75b01070/main.gif',
    },
    id: '1',
    mainImage: 'https://i2.au.reastatic.net/640x480/20bfc8668a30e8cabf045a1cd54814a9042fc715a8be683ba196898333d68cec/main.jpg',
  },

];

const mockSavedData = [
  {
    price: '$526,500',
    agency: {
      brandingColors: {
        primary: '#000000',
      },
      logo: 'http://i2.au.reastatic.net/agencylogo/WVYSSK/2/20140701084436.gif',
    },
    id: '4',
    mainImage:
      'http://i2.au.reastatic.net/640x480/5e84d96722dda3ea2a084d6935677f64872d1d760562d530c3cabfcb7bcda9c2/main.jpg',
  },
];

const server = setupServer(
  rest.get(settings.BASE_API_DOMAIN, (req, res, ctx) => res(ctx.json(mockResultsData))),
);

beforeAll(() => server.listen());
afterEach(() => {
  server.resetHandlers();
  cleanup();
});
afterAll(() => server.close());

describe('Render Test', () => {
  it('1: expect rendering correct when type=results', () => {
    const { container } = render(
      <redux.Provider store={store}>
        <PropertyList type="results" />
      </redux.Provider>,
    );
    expect(container).toMatchSnapshot();
  });

  it('2: expect rendering correct when type=saved', () => {
    const { container } = render(
      <redux.Provider store={store}>
        <PropertyList type="saved" />
      </redux.Provider>,
    );
    expect(container).toMatchSnapshot();
  });
});

describe('Function Test', () => {
  it('1: expect getResultsAction becalled', async () => {
    const { getByTestId } = render(
      <redux.Provider store={store}>
        <PropertyList type="results" />
      </redux.Provider>,
    );

    await waitFor(() => getByTestId('property'));

    expect(getByTestId('property')).toBeInTheDocument();
  });

  it('2: expect getSavedAction becalled', async () => {
    server.use(
      rest.get(settings.SAVED_BASE_API_DOMAIN, (req, res, ctx) => res(ctx.json(mockSavedData))),
    );

    const { getByTestId } = render(
      <redux.Provider store={store}>
        <PropertyList type="saved" />
      </redux.Provider>,
    );

    await waitFor(() => getByTestId('property'));

    expect(getByTestId('property')).toBeInTheDocument();
  });

  it('3: expect results button onclick event', async () => {
    const resultScreen = render(
      <redux.Provider store={store}>
        <PropertyList type="results" />
      </redux.Provider>,
    );

    const savedScreen = render(
      <redux.Provider store={store}>
        <PropertyList type="saved" />
      </redux.Provider>,
    );
    await waitFor(() => resultScreen.getByText('$726,500'));
    fireEvent.click(resultScreen.getByTestId('testresults'));
    expect(savedScreen.getByText('$726,500')).toBeTruthy();
  });

  it('4: expect saved button onclick event', async () => {
    server.use(
      rest.get(settings.SAVED_BASE_API_DOMAIN, (req, res, ctx) => res(ctx.json(mockSavedData))),
    );

    const savedScreen = render(
      <redux.Provider store={store}>
        <PropertyList type="saved" />
      </redux.Provider>,
    );
    await waitFor(() => savedScreen.getByText('$526,500'));
    fireEvent.click(savedScreen.getByTestId('testsaved'));
    expect(savedScreen.container.querySelector('testsaved')).not.toBeInTheDocument();
  });

  it('5: expect get result failed', async () => {
    server.use(
      rest.get(settings.BASE_API_DOMAIN, (req, res, ctx) => res(ctx.status(500), ctx.json({ message: 'Internal Server Error' }))),
    );

    const { getByTestId } = render(
      <redux.Provider store={store}>
        <PropertyList type="results" />
      </redux.Provider>,
    );

    await waitFor(() => getByTestId('loading'));

    expect(getByTestId('loading')).toBeInTheDocument();
  });

  it('6: expect get saved failed', async () => {
    server.use(
      rest.get(settings.SAVED_BASE_API_DOMAIN, (req, res, ctx) => res(ctx.status(500), ctx.json({ message: 'Internal Server Error' }))),
    );

    const { getByTestId } = render(
      <redux.Provider store={store}>
        <PropertyList type="saved" />
      </redux.Provider>,
    );

    await waitFor(() => getByTestId('loading'));

    expect(getByTestId('loading')).toBeInTheDocument();
  });

  it('7: expect get repeat', async () => {
    server.use(
      rest.get(settings.SAVED_BASE_API_DOMAIN, (req, res, ctx) => res(ctx.json(mockResultsData))),
    );

    const resultScreen = render(
      <redux.Provider store={store}>
        <PropertyList type="results" />
      </redux.Provider>,
    );

    const savedScreen = render(
      <redux.Provider store={store}>
        <PropertyList type="saved" />
      </redux.Provider>,
    );
    await waitFor(() => resultScreen.getByTestId('testresults'));
    fireEvent.click(resultScreen.getByTestId('testresults'));
    expect(savedScreen.container.firstChild.childNodes).toHaveLength(4);
  });
});
