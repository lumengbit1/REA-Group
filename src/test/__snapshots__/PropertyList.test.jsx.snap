// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Test 1: expect rendering correct when type=results 1`] = `<div />`;

exports[`Render Test 2: expect rendering correct when type=saved 1`] = `<div />`;
